src/mysql_protocol.erl

```diff
@@ -1230,37 +1294,93 @@ get_null_terminated_binary(<<0, Rest/binary>>, Acc) ->
+    %% (transcribed):
+    %% SHA256( password ) XOR SHA256( SHA256( SHA256( password ) ) <concat>
+    %%                                        "20-bytes random data from server" )
+    Salt1 = trim_salt(Salt),
+    <<Hash1Num:256>> = Hash1 = crypto:hash(sha256, Password),
+    Hash2 = crypto:hash(sha256, Hash1),
+    <<Hash3Num:256>> = crypto:hash(sha256, <<Hash2/binary, Salt1/binary>>),
+    <<(Hash1Num bxor Hash3Num):256>>;
+hash_password(?authmethod_sha256_password, Password, Salt) ->
+    %% sha256_password authentication is superseded by
+    %% caching_sha2_password.
+    hash_password(?authmethod_caching_sha2_password, Password, Salt);
+hash_password(UnknownAuthMethod, _, _) ->
+    error({auth_method, UnknownAuthMethod}).
+
+encrypt_password(AuthMethod, Password, Salt, PubKey, ServerVersion)
+  when not is_binary(Password) ->
+    encrypt_password(AuthMethod, iolist_to_binary(Password), Salt, PubKey,
+                     ServerVersion);
+encrypt_password(?authmethod_caching_sha2_password, Password, Salt, PubKey,
+                 ServerVersion) ->
+    %% From http://www.dataarchitect.cloud/preparing-your-community-connector-for-mysql-8-part-2-sha256/:
+    %% "The password is "obfuscated" first by employing a rotating "xor" against
+    %% the seed bytes that were given to the authentication plugin upon initial
+    %% handshake [the auth plugin data].
+    %% [...]
+    %% Buffer would then be encrypted using the RSA public key the server passed
+    %% to the client.  The resulting buffer would then be passed back to the
+    %% server."
+    Salt1 = trim_salt(Salt),
+
+    %% While the article does not mention it, the password must be null-terminated
+    %% before obfuscation.
+    Password1 = <<Password/binary, 0>>,
+    Salt2 = case byte_size(Salt1)<byte_size(Password1) of
+        true ->
+            binary:copy(Salt1, (byte_size(Password1) div byte_size(Salt1)) + 1);
+        false ->
+            Salt1
+    end,
+    Size = bit_size(Password1),
+    <<PasswordNum:Size>> = Password1,
+    <<SaltNum:Size, _/bitstring>> = Salt2,
+    Password2 = <<(PasswordNum bxor SaltNum):Size>>,
+
+    %% From http://www.dataarchitect.cloud/preparing-your-community-connector-for-mysql-8-part-2-sha256/:
+    %% "It's important to note that a incompatible change happened in server 8.0.5.
+    %% Prior to server 8.0.5 the encryption was done using RSA_PKCS1_PADDING.
+    %% With 8.0.5 it is done with RSA_PKCS1_OAEP_PADDING."
+    RsaPadding = case ServerVersion < [8, 0, 5] of
+        true -> rsa_pkcs1_padding;
+        false -> rsa_pkcs1_oaep_padding
+    end,
+    public_key:encrypt_public(Password2, PubKey, [{rsa_padding, RsaPadding}]);
+encrypt_password(?authmethod_sha256_password, Password, Salt, PubKey,
+                 ServerVersion) ->
+    %% sha256_password authentication is superseded by
+    %% caching_sha2_password.
+    encrypt_password(?authmethod_caching_sha2_password, Password, Salt, PubKey,
```

Comment:
Why does this function even have AuthMethod as a parameter? It is only used for these two and they are using the same algorithm. Maybe the AuthMethod parameter can be removed.

Reply:
Wasn't sure here. I included it preemptively in case more auth methods show up. OTOH, I can remove it for now and we can add it back when new ones show up. What do you say?

Reply:
Remove it then. Preemptively adding stuff is rarily good. "Ain't gonna need it" as they say.